### PR TITLE
BUPH-103 | per file template ordering

### PR DESCRIPTION
**Pull request by jpmarcotte**

### CODE REVIEW

File: src/V1/Actor/Template/Builder.php [hunk 1]
```diff
@@ -54,16 +56,37 @@ private function buildActorTemplateFilePathCandidates(): array
...
         $templateTreeMap = $this->getTemplateTreeMapRepository()->get();
-        foreach($templateTreeMap as $templateTree) {
-            $actorTemplateFilePathCandidate = sprintf(
-                '%s/%s',
-                $templateTree->getDirectoryPath(),
-                $this->getFabricationFileActor()->getTemplateRelativeFilePath()
-            );
 
-            $actorTemplateFilePathCandidates[] = $actorTemplateFilePathCandidate;
+        // Let the Fabrication File preferences be used over global ones
```
Comment (rhift) [inline src/V1/Actor/Template/Builder.php]: Method: "If the Fabrication File specifies a list of Preferred Template Tree directory paths by using their associated names I collect those in an array in order first, using their paths as keys and concatenating the Template's relative file path to them by asking buildTemplatePath to do that for me, then I add those specified globally to the same array, in the same manner. The way PHP arrays work will deduplicate the globally specified paths, preserving global ordering and existence when not usurped by a Fabrication File preference. If a Preferred Template Tree does not specify a name that exists in the global specification, I throw an exception. I return an array of possible Template file paths. I assume another Actor will validate if they exist on the FS."

File: src/V1/FabricationFile.php [hunk 2]
```diff
@@ -155,4 +156,29 @@ public function setDirectoryPath(string $DirectoryPath): FabricationFileInterfac
...
+
+        return $this->PreferredTemplateTrees;
+    }
+
+    public function hasPreferredTemplateTrees(): bool
+    {
+        return $this->PreferredTemplateTrees !== null;
+    }
+
+    public function setPreferredTemplateTrees(string ...$PreferredTemplateTrees): FabricationFileInterface
```
Comment (rhift) [inline src/V1/FabricationFile.php]: consider modeling Preferred Template Trees as a first class DAO

File: src/V1/FabricationFile/Builder.php [hunk 3]
```diff
@@ -20,13 +20,21 @@ class Builder implements BuilderInterface
...
         $fabricationFile->setFileName($this->getFileName());
         $fabricationFile->setFilePath($this->getSplFileInfo()->getPathname());
         $fabricationFile->setDirectoryPath($this->getSplFileInfo()->getPath());
         $fabricationFile->setRelativeDirectoryPath($this->getSplFileInfo()->getRelativePath());
         $fabricationFile->setRelativeFilePath($this->getSplFileInfo()->getRelativePathname());
+
+        if (isset($fabricationFileContents[FabricationFileInterface::KEY_PREFERRED_TEMPLATE_TREES])) {
+            $fabricationFile->setPreferredTemplateTrees(
+                ...$fabricationFileContents[FabricationFileInterface::KEY_PREFERRED_TEMPLATE_TREES]
+            );
```
Comment (rhift) [inline src/V1/FabricationFile/Builder.php]: nice

File: tests/v1/PerFabletTemplateTrees/src/FabletDefinedTrees.buphalo.v1.fabrication.yml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+preferred_template_trees:
```
Comment (rhift) [inline tests/v1/PerFabletTemplateTrees/src/FabletDefinedTrees.buphalo.v1.fabrication.yml]: nice naming choice for the public API =)

File: src/V1/FabricationFile.php (hunk 2 above)
Comment (jpmarcotte) [reply src/V1/FabricationFile.php]: Hrmm... I think I have some thoughts on how to do that, but I might tackle that as as a follow-up refactor to the full thing if you're cool with that.

Comment (rhift) [reply src/V1/FabricationFile.php]: Totally cool with that

Comment (jpmarcotte) [reply src/V1/FabricationFile.php]: Great. Created [BUPH-113](https://55places.atlassian.net/browse/BUPH-113)

**Pull request merged**